File: n8n-plugin-worker/app/sandbox-client.mjs
/** 通过来源专用 Unix Socket 调用插件 Docker 沙箱 Broker。 */

import { request as httpRequest } from 'node:http'

export class SandboxError extends Error {
  /** 保存有限状态码和稳定错误码。 */
  constructor(status, code) {
    const normalized = /^[A-Z0-9_]{1,80}$/.test(String(code || '')) ? String(code) : 'PLUGIN_SANDBOX_UNAVAILABLE'
    super(normalized)
    this.status = Number(status) >= 400 && Number(status) <= 599 ? Number(status) : 503
    this.code = normalized
  }
}

export class SandboxClient {
  /** 固定 Unix Socket、调用超时和响应上限。 */
  constructor(socketPath) {
    this.socketPath = socketPath
    this.timeout = Math.max(1000, Math.min(Number(process.env.PLUGIN_WORKER_TIMEOUT_SECONDS || 660) * 1000, 660000))
    this.maximum = Math.max(1024, Math.min(Number(process.env.PLUGIN_WORKER_MAX_RESPONSE_BYTES || 16 * 1024 * 1024),
      16 * 1024 * 1024))
  }

  /** 发送固定操作的有限 JSON，并拒绝网络回退和畸形响应。 */
  async request(operation, value) {
    if (!['inspect', 'invoke', 'remove'].includes(operation)) throw new SandboxError(400, 'PLUGIN_SANDBOX_OPERATION_INVALID')
    const body = Buffer.from(JSON.stringify(value))
    return new Promise((resolve, reject) => {
      const request = httpRequest({ socketPath: this.socketPath, path: `/sandbox/${operation}`, method: 'POST',
        headers: { 'content-type': 'application/json', 'content-length': body.length } }, response => {
        const chunks = []; let size = 0
        response.on('data', chunk => {
          size += chunk.length
          if (size > this.maximum) response.destroy(new SandboxError(502, 'PLUGIN_SANDBOX_OUTPUT_INVALID'))
          else chunks.push(chunk)
        })
        response.once('error', error => reject(error instanceof SandboxError ? error
          : new SandboxError(503, 'PLUGIN_SANDBOX_UNAVAILABLE')))
        response.once('end', () => {
          try {
            const result = JSON.parse(Buffer.concat(chunks).toString('utf8'))
            if (!result || typeof result !== 'object' || Array.isArray(result)) throw new Error('invalid')
            if (Math.floor((response.statusCode || 500) / 100) !== 2) {
              reject(new SandboxError(response.statusCode, result.error)); return
            }
            resolve(result)
          } catch { reject(new SandboxError(502, 'PLUGIN_SANDBOX_OUTPUT_INVALID')) }
        })
      })
      request.setTimeout(this.timeout, () => request.destroy(new SandboxError(504, 'PLUGIN_SANDBOX_TIMEOUT')))
      request.once('error', error => reject(error instanceof SandboxError ? error
        : new SandboxError(503, 'PLUGIN_SANDBOX_UNAVAILABLE')))
      request.end(body)
    })
  }

  /** 验证来源专用 Broker Socket 及其 Docker Engine 可用。 */
  async healthy() {
    return new Promise(resolve => {
      const request = httpRequest({ socketPath: this.socketPath, path: '/health', method: 'GET' }, response => {
        const chunks = []
        response.on('data', chunk => chunks.push(chunk))
        response.on('end', () => {
          try { resolve(response.statusCode === 200 && JSON.parse(Buffer.concat(chunks).toString('utf8')).status === 'UP') }
          catch { resolve(false) }
        })
      })
      request.setTimeout(3000, () => request.destroy())
      request.on('error', () => resolve(false))
      request.end()
    })
  }
}

File: dify-plugin-worker/app/sandbox_client.py
"""通过来源专用 Unix Socket 调用插件 Docker 沙箱 Broker。"""

from __future__ import annotations

import http.client
import json
import os
import socket
from typing import Any


class SandboxError(RuntimeError):
    """表示 Broker 拒绝、超时或无法完成一次沙箱操作。"""

    def __init__(self, status: int, code: str) -> None:
        """保存有限 HTTP 状态和稳定错误码。"""
        super().__init__(code)
        self.status = status if 400 <= status <= 599 else 503
        self.code = code if code.replace("_", "").isalnum() and code.upper() == code else "PLUGIN_SANDBOX_UNAVAILABLE"


class UnixHTTPConnection(http.client.HTTPConnection):
    """把标准库 HTTP 客户端固定到单个 Unix Socket。"""

    def __init__(self, socket_path: str, timeout: int) -> None:
        """记录不可由请求修改的 Socket 路径。"""
        super().__init__("plugin-sandbox", timeout=timeout)
        self.socket_path = socket_path

    def connect(self) -> None:
        """只建立本机 Unix 域连接，不允许网络回退。"""
        connection = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        connection.settimeout(self.timeout)
        connection.connect(self.socket_path)
        self.sock = connection


class SandboxClient:
    """向 Broker 发送有限 JSON，并限制响应体和错误细节。"""

    def __init__(self, socket_path: str) -> None:
        """读取调用超时与响应上限。"""
        self.socket_path = socket_path
        self.timeout = max(1, min(int(os.getenv("PLUGIN_WORKER_TIMEOUT_SECONDS", "660")), 660))
        self.maximum = max(1024, min(int(os.getenv("PLUGIN_WORKER_MAX_RESPONSE_BYTES", str(16 * 1024 * 1024))),
                                     16 * 1024 * 1024))

    def request(self, operation: str, value: dict[str, Any]) -> dict[str, Any]:
        """调用固定操作并解析 Broker 返回的单个 JSON 对象。"""
        if operation not in {"inspect", "invoke", "remove"}:
            raise SandboxError(400, "PLUGIN_SANDBOX_OPERATION_INVALID")
        body = json.dumps(value, ensure_ascii=False, separators=(",", ":")).encode("utf-8")
        connection = UnixHTTPConnection(self.socket_path, self.timeout)
        try:
            connection.request("POST", f"/sandbox/{operation}", body,
                               {"Content-Type": "application/json", "Content-Length": str(len(body))})
            response = connection.getresponse()
            payload = response.read(self.maximum + 1)
            if len(payload) > self.maximum:
                raise SandboxError(502, "PLUGIN_SANDBOX_OUTPUT_INVALID")
            result = json.loads(payload)
            if not isinstance(result, dict):
                raise SandboxError(502, "PLUGIN_SANDBOX_OUTPUT_INVALID")
            if response.status // 100 != 2:
                raise SandboxError(response.status, str(result.get("error", "PLUGIN_SANDBOX_UNAVAILABLE"))[:80])
            return result
        except SandboxError:
            raise
        except (OSError, ValueError, json.JSONDecodeError) as exception:
            raise SandboxError(503, "PLUGIN_SANDBOX_UNAVAILABLE") from exception
        finally:
            connection.close()

    def healthy(self) -> bool:
        """验证来源专用 Broker Socket 及其 Docker Engine 可用。"""
        connection = UnixHTTPConnection(self.socket_path, min(self.timeout, 3))
        try:
            connection.request("GET", "/health")
            response = connection.getresponse()
            payload = response.read(1024)
            return response.status == 200 and json.loads(payload).get("status") == "UP"
        except (OSError, ValueError, json.JSONDecodeError):
            return False
        finally:
            connection.close()
